---
title: "A1/A2 Recap"
output:
  html_document:
    df_print: paged
---
## A1
### Library Preperation

We will start by checking if the required libraries and files are installed.
Then we will load the libraries using library().
```{r message=FALSE, warning=TRUE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")
}

if (!requireNamespace("biomaRt", quietly = TRUE)){
  BiocManager::install("biomaRt")
}

if (!requireNamespace("edgeR", quietly = TRUE)){
  BiocManager::install("edgeR")
}

library(BiocManager)
library(edgeR)
library(GEOmetadb)
library(biomaRt)
library(knitr)

if (!file.exists('GEOmetadb.sqlite')){
  getSQLiteFile()
}
```

### Data Selection

Now with all libraries installed, we will use the following commands to query the GEOmetadb (GEO database) to look for a dataset with a "raw counts" txt-supplmentary file. Note we are specifically querying datasets with submission dates > 2015-01-01, related to cancer and high throughput sequencing data of Homo sapiens. 
```{r, warning=FALSE, message=FALSE}
sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
             " gse.submission_date,",
             " gse.supplementary_file",
             "FROM",
             "  gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
             "  JOIN gpl ON gse_gpl.gpl=gpl.gpl",
             "WHERE",
             "  gse.submission_date > '2015-01-01' AND",
             "  gse.title LIKE '%cancer%' AND", 
             "  gpl.organism LIKE '%Homo sapiens%' AND",
             "  gpl.technology LIKE '%high-throughput seq%' ",
             "  ORDER BY gse.submission_date DESC",sep=" ")
con <- dbConnect(SQLite(), 'GEOmetadb.sqlite')
rs <- dbGetQuery(con, sql)
unlist(lapply(rs$supplementary_file,
              FUN = function(x){x <- unlist(strsplit(x,";")) ;
              x <- x[grep(x,pattern="txt",ignore.case = TRUE)];
              tail(unlist(strsplit(x,"/")),n=1)})) [1:30]
```

The chosen dataset is "GSE164531" which is an RNA-seq dataset using the Illumina Hiseq 2000 platform for identification of NEDD9 regulated genes in VCaP prosatate cancer cells. We download the counts data using the following commands and check the dimensions of the data.
```{r include=FALSE, warning=FALSE, message=FALSE}
sfiles <- getGEOSuppFiles('GSE164531')
fnames <- rownames(sfiles)
```

We then load the counts data into R and check the dimensions.
```{r, warning=FALSE, message=FALSE}
data <- read.delim(fnames[1], header = TRUE, check.names = FALSE)
dim(data)
```

Then we quickly check the head of the data (first 10 rows) to explore further.
```{r, warning=FALSE, message=FALSE}
kable(head(data), format = 'html', caption = "Table 0.1: GSE164531 Dataset")
```

### Initial Exploratory Analysis

#### Duplicate Genes

The first exploratory analysis is checking for duplicate genes in our data. We run the following commands and find that there are no duplicate genes present in the data. Resulting in no further actions required to deal with duplicate genes.

```{r, warning=FALSE, message=FALSE}
summarized_gene_counts <- sort(table(data$EnsemblID),decreasing = TRUE)
kable(summarized_gene_counts[which(summarized_gene_counts > 1)[1:10]], format = 'html', caption = "Table 0.2: Duplicate genes and it's frequencies")
```

#### Missing Data

Now we check for any missing data in the dataset. 

```{r, warning=FALSE, message=FALSE}
# Check which rows have missing data
which(!complete.cases(data))
```

We see that row 58736 has missing data, we access this row to explore further.

```{r, warning=FALSE, message=FALSE}
data[58736,]
```
Seems like this row is just an empty row with no meaningful data. As it contains no data, we can safely remove the row without affecting the data-set (We remove the row as it N/A values may interfere with next steps such as normalization).
```{r, warning=FALSE, message=FALSE}
data <- data[complete.cases(data),]
dim(data)
```

#### Grouping the Data
We will define groups for the data according to each biological condition (experiment design) for later steps such as normalization.
```{r, warning=FALSE, message=FALSE}
samples <- data.frame(lapply(colnames(data)[3:8], FUN=function(x){unlist(strsplit(x, split = "_"))[c(1, 2)]}))
colnames(samples) <- colnames(data)[3:8]
rownames(samples) <- c("condition", "replicate")
samples <- data.frame(t(samples))
samples
```

### Filtering Data
We now filter out low counts from the data according to the edgeR protocol. The threshold is set to 3 because edgeR recommends the threshold to be the number of replications which in our data set is 3 (Would be interesting to see what happens if we change our threshold, but we will keep it to 3 for now).
```{r, warning=FALSE, message=FALSE}
cpms <- edgeR::cpm(data[, 3:8])
rownames(cpms) <- data[, 1]
# Threshold set to 3 as recommended by edgeR protocol
keep <- rowSums(cpms > 1) >= 3
dataFiltered <- data[keep, ]
head(dataFiltered)
```

We quickly compare the dimensions of the filtered data and original data.
```{r, warning=FALSE, message=FALSE}
dim(data)
```

```{r, warning=FALSE, message=FALSE}
kable(head(dataFiltered), format = 'html', caption = "Table 0.3: Filtered Data")
```
We notice that the total number of genes reduced to 14682 from 58735 after filtering out low expression data using the edgeR filtering protocol.


### Mapping Data

Now we will map the filtered data to HUGO gene symbols using grch38.p13 and the biomaRt package. The following commands creates a ensembl_gene_id and HUGO gene symbol conversion table which we will use to merge with the expression data after normalization.
```{r, warning=FALSE, message=FALSE}
ensembl <- useEnsembl("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
conversionStash <- "data_conversion.rds"
if(file.exists(conversionStash)){
  dataIdConversion <- readRDS(conversionStash)
} else {
  dataIdConversion <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = dataFiltered$EnsemblID,
                               mart = ensembl)
  saveRDS(dataIdConversion, conversionStash)
}

kable(head(dataIdConversion), format = 'html', caption = "Table 0.4: Ensembl Gene Id to hgnc symbol")
```

### Normalization

We will now normalize our data using the TMM method using the following commands.
```{r, warning=FALSE, message=FALSE}
filteredDataMatrix <- as.matrix(dataFiltered[,3:8])
rownames(filteredDataMatrix) <- dataFiltered$EnsemblID
d = DGEList(counts = filteredDataMatrix, group = samples$condition)
d = calcNormFactors(d)
normalizedCounts <- cpm(d)
```

Now we will merge the normalizedCounts with our identifiers that we mapped in our previous step. Now we have a dataFrame with mapped ensembl_gene_id, HUGO gene symbols and normalized expression data.

```{r, warning=FALSE, message=FALSE}
dataFilteredAnnot <- merge(dataIdConversion, normalizedCounts, by.x = 1, by.y = 0, all.y=TRUE)
kable(dataFilteredAnnot[1:5,1:8],type = "html", caption = "Table 0.5: Normalized Data")
```


#### Pre-normalization Density vs Post-normalization Density
The density plots for pre-normalization and post-normalization data is plotted for comparison.

##### Pre-normalization:
```{r, warning=FALSE, message=FALSE, fig.cap= "Fig. 0.1: Density Plot for Pre-normalization data"}
pre_data2plot <- log2(cpm(dataFiltered[, 3:8]))
pre_counts_density <- apply(log2(cpm(dataFiltered[, 3:8])), 2, density)
xlim <- 0
ylim <- 0
for (i in 1:length(pre_counts_density)) {
  xlim <- range(c(xlim, pre_counts_density[[i]]$x)); 
  ylim <- range(c(ylim, pre_counts_density[[i]]$y))
  }
cols <- rainbow(length(pre_counts_density))
ltys <- rep(1, length(pre_counts_density))
#plot the first density plot to initialize the plot
plot(pre_counts_density[[1]], xlim=xlim, ylim=ylim, type="n", ylab="Smoothing density of log2-CPM", main="", cex.lab = 0.85)
#plot each line
for (i in 1:length(pre_counts_density)) lines(pre_counts_density[[i]], col=cols[i], lty=ltys[i])
#create legend
legend("topright", colnames(pre_data2plot),  
       col=cols, lty=ltys, cex=0.75,
       border ="blue",  text.col = "green4",
       merge = TRUE, bg = "gray90")
```

<br>

##### Post-normalization:
```{r, warning=FALSE, message=FALSE, fig.cap = "Fig. 0.2: Density Plot for Post-normalization data"}
post_data2plot <- log2(cpm(dataFilteredAnnot[, 3:8]))
post_counts_density <- apply(log2(cpm(dataFilteredAnnot[, 3:8])), 2, density)
xlim <- 0
ylim <- 0
for (i in 1:length(post_counts_density)) {
  xlim <- range(c(xlim, post_counts_density[[i]]$x)) 
  ylim <- range(c(ylim, post_counts_density[[i]]$y))
  }
cols <- rainbow(length(post_counts_density))
ltys <- rep(1, length(post_counts_density))
#plot the first density plot to initialize the plot
plot(post_counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
ylab="Smoothing density of log2-CPM", main="", cex.lab = 0.85)
#plot each line
for (i in 1:length(post_counts_density)) lines(post_counts_density[[i]], col=cols[i], lty=ltys[i])
#create legend
legend("topright", colnames(post_data2plot),  
       col=cols, lty=ltys, cex=0.75, 
       border ="blue",  text.col = "green4", 
       merge = TRUE, bg = "gray90")
```

<br>
Notice no significance difference between pre and post normalization. I believe this is due to the fact that my original data happens to be close to a normal distribution, thus normalization with TMM method did not make any significant changes.

### Missing Identifiers

We will now deal with missing identifiers (missing HUGO gene symbols). First, we will check how many missing identifiers are present in our data right now.

```{r, warning=FALSE, message=FALSE}
ensembl_id_missing_gene <- dataFilteredAnnot$ensembl_gene_id[which((dataFilteredAnnot$hgnc_symbol) == "")]
length(ensembl_id_missing_gene)
```
We see that we have 732 missing identifiers, lets check out some of these rows with missing identifiers.
```{r, warning=FALSE, message=FALSE}
kable(dataFilteredAnnot[which((dataFilteredAnnot$hgnc_symbol == ""))[1:5],1:8], type="html", caption = "Table 0.6: Genes with missing identifiers")
```
As the missing identifiers are only ~5% of the data set, we will keep these rows for now and continue with our analysis. Also, we have knowledge of the rows with missing identifiers which will allow us to remove/modify how we handle these missing identifiers easily later on in our analysis and removing these data at this step does not seem the best practice (would be interesting to see how these genes with missing identifiers result in our final analysis).

### Final Data

Now we present the final data set that has been filtered, normalized and mapped to HUGO gene symbols with 14683 genes (rows).
```{r, warning=FALSE, message=FALSE}
kable(head(dataFilteredAnnot), format = "html", caption = "Table 0.7: Final Dataset")
```

```{r, warning=FALSE, message=FALSE}
dim(dataFilteredAnnot)
```


## A2

### Differential Gene Expression

#### Adjustments from A1

##### Duplicate Genes
First we will remove duplicate genes as I have made a mistake in this step for A1. We see that there are two duplicate gene_ids each with a frequency of 2.

```{r, warning=FALSE, message=FALSE}
summarizedGeneCounts <- sort(table(dataFilteredAnnot$ensembl_gene_id), decreasing = TRUE)
kable(summarizedGeneCounts[which(summarizedGeneCounts > 1)[1:5]], format = 'html', caption = "Table 1: Duplicate Genes and it's Frequencies.")
```

<br>

We will then remove 1 row for each duplicate ID as it contains exactly the same data.

```{r, warning=FALSE, message=FALSE}
dataFilteredAnnot <- dataFilteredAnnot[-c(12987, 13672),]
```

##### MDS Plot
Then we will check the MDS plots to decide factors in our model as I have not included a MDS plot in A1.

The MDS plot colored by condition (red for shNEDD9 and blue for shNTC) show that expression depends heavily on the experiment condition (execpt shNTC_2 which seems to be an outlier).

```{r, warning=FALSE, message=FALSE, fig.cap= "Fig. 1: MDS plot coloured by experimental condition (shNEDD9 vs shNTC)"}
plotMDS(d, labels = rownames(samples), col = c("red", "blue")[factor(samples$condition)])
legend(-1.5, 0.4, legend=c("shNEDD9", "shNTC"), fill = c("red", "blue"), cex = 0.6)
```
<br>

As we can see the groups are clustered based on cell type except shNTC_2, we will create a design matrix with samples$condition. Then we will create an expression matrix.

#### Calculate P-value: Limma Method

We first design a model with only condition as factor since our MDS plot shows that condition is the most significant factor.

```{r, warning=FALSE, message=FALSE}
modelDesign <- model.matrix( ~ samples$condition)
kable(modelDesign[1:6,], type="html", caption = "Table 2: Model Design Matrix with only condition as factor")
```

<br>

Then we create an expression matrix and fit our model to get our p-values as shown in the below table. 
```{r, warning=FALSE, message=FALSE}
library(limma)
expressionMatrix <- as.matrix(dataFilteredAnnot[, 3:8])
rownames(expressionMatrix) <- dataFilteredAnnot$ensembl_gene_id
colnames(expressionMatrix) <- colnames(dataFilteredAnnot[3:8])
minimalSet <- ExpressionSet(assayData = expressionMatrix)
fit <- lmFit(minimalSet, modelDesign)
fit2 <- eBayes(fit, trend = TRUE)
topfit <- topTable(fit2, coef = ncol(modelDesign), adjust.method = "BH", number = nrow(expressionMatrix))
outputHits <- merge(dataFilteredAnnot[, 1:2], topfit, by.y= 0, by.x = 1, all.y = TRUE)
kable(outputHits[1:10, 1:8], type = "html", row.names = FALSE, caption = "Table 3: Results from Limma method for differential analysis")
```
<br>

From the limma method, we see that 10108 genes are significantly deferentially expressed with p-value threshold of 0.05. The threshold of 0.05 was chosen because it is considered the standard (to my knowledge). Eventhough we ended up with more genes passing than expected with the 0.05 p-value threshold, we are keeping the threshold to avoid a form of "p-hacking" (changing the threshold only to reduce the number of genes passing). 
```{r, warning=FALSE, message=FALSE}
length(which(outputHits$P.Value < 0.05))
```

<br>

We also see 9728 genes passing correction using the Benjamni - hochberg method with threshold of 0.05. The Benjamni - hochberg method was chosen as I remember it being mentioned as a standard method for RNA seq differential expression analysis from class (correct me if I am wrong here). Also, the threshold of 0.05 was chosen again to be consistant and seems to be the standard threshold in most cases. 
```{r, warning=FALSE, message=FALSE}
length(which(outputHits$adj.P.Val < 0.05))
```


#### Calculate P-value: EdgeR Quasi liklihood Method

Now we will use the Quasi likelihood of the EdgeR package to calculate p-values as the following:

```{r, warning=FALSE, message=FALSE}
d <- DGEList(counts=expressionMatrix, group=samples$condition)
d <- estimateDisp(d, modelDesign)
fit <- glmQLFit(d, modelDesign)
r <- glmQLFTest(fit)
qlfOutputHits <- topTags(r, n = nrow(dataFilteredAnnot), adjust.method = "BH")
kable(qlfOutputHits$table[1:10, 1:5], type = "html", row.names = FALSE, caption = "Table 4: Results from Quasi liklihood method for differential analysis")
```

From the Quasi likelihood method, we see that 6336 genes are significantly deferentially expressed with p-value threshold of 0.05. Again, we are using a p-value threshold of 0.05 as it seems to be the standard for these differential analysis.
```{r, warning=FALSE, message=FALSE}
length(which(qlfOutputHits$table$PValue < 0.05))
```
<br>

We also see 5354 genes passing correction using the Benjamni - hochberg method with threshold of 0.05. Same reason as the limma method, threshold of 0.05 and Quasi likelihood method were selected as these two were considered "standards" during class.
```{r, warning=FALSE, message=FALSE}
length(which(qlfOutputHits$table$FDR < 0.05))
```

<br>

#### Volcano Plot of Differentially Expressed Genes (From Quasi Likelihood Method)

We will continue our analysis from results of the Quasi Likelihood Method as "Limma guide direct users to use edgeR up to the point of calculating
differential expression." I purposely went through the Limma method earlier to compare results and practice the process of using Limma, but now we will only consider results from the Quasi Likelihood method for the rest of our analysis.

<br>

Now we will create a volcano plot to show the amounts of differentially expressed genes. Non-signiciant genes in grey, up regualated genes in red and down regualated genes in blue. The gene of interest, NEDD9, is colored black to highlight. 

```{r, warning=FALSE, message=FALSE, fig.cap= "Fig. 2: Volcano Plot of differentially expressed genes with FDR"}
col <- vector(mode="character", length = nrow(dataFilteredAnnot))
for (i in 1:nrow(qlfOutputHits$table)) {
  if (qlfOutputHits$table$logFC[i] < 0 && qlfOutputHits$table$FDR[i] < 0.05) {
    col[i] <- "blue"
  } else if (qlfOutputHits$table$logFC[i] > 0 && qlfOutputHits$table$FDR[i] < 0.05) {
    col[i] <- "red"
  } else {
    col[i] <- "grey"
  }
}
  
col[which(row.names(qlfOutputHits$table) == "ENSG00000111859")] <- "black"

plot(qlfOutputHits$table$logFC,
     -log(qlfOutputHits$table$FDR, base=10),
     col = col,
     xlab = "logFC",
     ylab ="-log(FDR)", 
     main="Volcano Plot of Differentially Expressed Genes")

legend(-6.9, 68, legend=c("Up Regulated Genes", "Down Regulated genes", "Non-significant", "NEDD9"), fill = c("blue", "red", "grey", "black"), cex = 0.5)
```
<br>

**Note that NEDD9 is in the up-regulated gene because I ran the analysis backwards NEDD9 silenced (shNEDD9 condition) vs Control (shNTC condition) instead of control vs silenced.**

<br>

#### Heatmap of Top Hits

We will portray a heatmap of the top hits with the following:

```{r, warning=FALSE, message=FALSE, fig.cap= "Fig. 3: Heatmap of Top Hits"}
if(!requireNamespace("circlize", quietly = TRUE)) {
  install.packages("circlize")
}

if(!requireNamespace("ComplexHeatmap", quietly = TRUE)) {
  BiocManager::install("ComplexHeatmap")
}

library(circlize)
library(ComplexHeatmap)

hmMatrix <- dataFilteredAnnot[, 3:ncol(dataFilteredAnnot)]
rownames(hmMatrix) <- dataFilteredAnnot$ensembl_gene_id

topHits <- rownames(qlfOutputHits$table)[qlfOutputHits$table$FDR < 0.05]
hmTopHits <- t(scale(t(hmMatrix[which(rownames(hmMatrix) %in% topHits), ])))

if (min(hmTopHits) == 0){
    hmCol = colorRamp2(c( 0, max(hmTopHits)), 
                             c( "white", "red"))
  } else {
    hmcol = colorRamp2(c(min(hmTopHits), 0, max(hmTopHits)), c("blue", "white", "red"))
  }

Heatmap(as.matrix(hmTopHits),
        cluster_rows = TRUE, show_row_dend = TRUE,
        cluster_columns = FALSE, show_column_dend = FALSE,
        col=hmcol, show_column_names = TRUE, 
        show_row_names = FALSE, show_heatmap_legend = TRUE, 
        column_title = "Heatmap of Top Hits")
```
<br>

From the heatmap, we can notice that the conditions do cluster together (shNEDD9 clustered on the left and shNTC clustered on the right). I believe this is due to the experimental design, when NEDD9 is suppressed the significantly differentiated genes will be up/down regulated in a similar way in the same NEDD9 suppressed condition (Same idea for the control, since NEDD9 is not suppressed, the genes up/down regulated will be similar for the control conditions).

<br>

### Thresholded Over-Representation Analysis

#### Creating Thresholded List

We will create a thresholded lists of genes.

```{r, warning=FALSE, message=FALSE}
upRegGenes <- row.names(qlfOutputHits$table)[which(qlfOutputHits$table$FDR < 0.05 & qlfOutputHits$table$logFC > 0)]
downRegGenes <- row.names(qlfOutputHits$table)[which(qlfOutputHits$table$FDR < 0.05 & qlfOutputHits$table$logFC < 0)]
allGenes <- c(upRegGenes, downRegGenes)
```

<br>

Now we save the list of genes in seperate files (table) for later access.
```{r, warning=FALSE, message=FALSE}
write.table(x=allGenes,
            "./allGenes.txt",sep='\t',
            row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=upRegGenes,
            "./upRegGenes.txt",sep='\t',
            row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=downRegGenes,
            "./downRegGenes.txt",sep='\t',
            row.names = FALSE,col.names = FALSE,quote = FALSE)
```

<br>

#### g:Profiler 

For over-representation analysis, we will use the g:Profiler method as it allows access to multiple annotation data sources such as Go biological process, Reactom and etc within the method. There is also a R package named "gprofiler2" which is a convenient way to run g:profiler within R (we will be doing this).

Let's begin with loading the "gprofiler2" library.

```{r, warning=FALSE, message=FALSE}
if (!requireNamespace("gprofiler2", quietly = TRUE)){
  install.packages("gprofiler2")
}
library(gprofiler2)
```

#### Annotation Data and versions

For annotation data, we will be using the Go biological process (GO-BP) and Reactome. GO-BP was chosen for the relevant biological processes of the significantly differentiated genes. Reactome was also chosen for relevent biological pathways of the significantly differentiated genes (Reactome and WikiPathways are both databases for biological pathways so I decided to only choose Reactome for this assignment). The combination of GO-BP and Reactome seems to be most reasonable giving annotation data of the biological processes and pathways.

<br>
The version for GO-BP:
```{r, warning=FALSE, message=FALSE}
version <- gprofiler2::get_version_info(organism = "hsapiens")
goBPVersion <- version$sources$`GO-BP`$version
goBPVersion
```

<br>

The version for Reactome:
```{r, warning=FALSE, message=FALSE}
reactomeVersion <- version$sources$REAC$version
reactomeVersion
```
<br>

From above, we can notice that the used GO-BP  version is the 2021-12-15 version. While the Reactome version is the 2022-01-03 version.

#### Running g:Profiler on all differentially expressed genes
We will first run a g:Profiler query for all significantly differentiated genes with the GO-BP and Reactome annotations. 

```{r, warning=FALSE, message=FALSE}
allGenesResults <- gost(
  allGenes,
  organism = "hsapiens",
  ordered_query = FALSE,
  multi_query = FALSE,
  significant = FALSE,
  exclude_iea = FALSE,
  measure_underrepresentation = FALSE,
  evcodes = FALSE,
  user_threshold = 0.05,
  correction_method = c("fdr"),
  sources = c("GO-BP", "REAC"),
  as_short_link = FALSE
)
```

<br>

Lets check how many genesets are returned with threshold of FDR < 0.05, we can see that there are 14565 gene sets returned from GO-BP and Reactome. 

```{r, warning=FALSE, message=FALSE}
length(allGenesResults$result$term_id)
```

<br>

Now, we can get the top GO-BP results for querying all genes:
```{r, warning=FALSE, message=FALSE}
kable(head(allGenesResults$result[, c("term_name", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 5: Top GO-BP Results for All Genes")
```
<br>

From the GO-BP results of all genes, it seems like the terms are related to cellular metabolic processes (macromolecules, protein and organonitrogen). Since the NEDD9 Gene is a member of the CRK-associated substrates family which play a role as a adhesion docking molecule, the top results of GO-BP (cellular metabolic processes) seems to make sense.

<br>

We also can get the top Reactome results for querying all genes:
```{r, warning=FALSE, message=FALSE}
kable(head(allGenesResults$result[allGenesResults$result$source == "REAC", c("term_name", "source", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 6: Top Reactome Results for All Genes")
```
<br>

From Entrex Gene Summary of NEDD9, it is known that NEDD9 plays a role in apoptosis and the cell cycle. Thus, the results from Reactome makes a lot of sense with the Cell Cycle term as the top result. I'm guessing the Cellular Response to stress/stimuli term is regarding apoptosis/cell cycle (but I am not too sure).

<br>

##### Running g:Profiler on Up-Regulated Genes

Now we will run the same query but only with the up-regulated genes.

```{r, warning=FALSE, message=FALSE}
upRegGenesResults <- gost(
  upRegGenes,
  organism = "hsapiens",
  ordered_query = FALSE,
  multi_query = FALSE,
  significant = FALSE,
  exclude_iea = FALSE,
  measure_underrepresentation = FALSE,
  evcodes = FALSE,
  user_threshold = 0.05,
  correction_method = c("fdr"),
  sources = c("GO-BP", "REAC"),
  as_short_link = FALSE
)
```

<br>

Lets check how many genesets are returned with threshold of FDR < 0.05, we can see that there are 12087 gene sets returned from GO-BP and Reactome. 
```{r, warning=FALSE, message=FALSE}
length(upRegGenesResults$result$term_id)
```


Now, we can get the top GO-BP results for querying up-regulated genes:
```{r, warning=FALSE, message=FALSE}
kable(head(upRegGenesResults$result[, c("term_name", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 7: Top GO-BP Results for Up-Regulated Genes")
```
<br>

The first thing we notice is the top terms are different from the top terms of the "all genes" GO-BP query. The up-regulated genes seem to be more focused on localization and development process.

<br>
We also can get the top Reactome results for querying up-regulated genes:
```{r, warning=FALSE, message=FALSE}
kable(head(upRegGenesResults$result[upRegGenesResults$result$source == "REAC", c("term_name", "source", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 8: Top Reactome Results for Up-Regulated Genes")
```
<br>
We also notice here that the top terms are different from the top terms of the "all genes" Reactome query. The up-regualted genes (genes more expressed in control since we did the analysis in reverse like mentioned earlier) seem to be more focused on membrane trafficking / signaling while "all genes" query is focusing more on cell cycle.
<br>

##### Running g:Profiler on Down-Regulated Genes

Now we will run the same query but only with the down-regulated genes.

```{r, warning=FALSE, message=FALSE}
downRegGenesResults <- gost(
  downRegGenes,
  organism = "hsapiens",
  ordered_query = FALSE,
  multi_query = FALSE,
  significant = FALSE,
  exclude_iea = FALSE,
  measure_underrepresentation = FALSE,
  evcodes = FALSE,
  user_threshold = 0.05,
  correction_method = c("fdr"),
  sources = c("GO-BP", "REAC"),
  as_short_link = FALSE
)
```

<br>

Lets check how many genesets are returned with threshold of FDR < 0.05, we can see that there are 11299 gene sets returned from GO-BP and Reactome. 

```{r, warning=FALSE, message=FALSE}
length(downRegGenesResults$result$term_id)
```

<br>

Now, we can get the top GO-BP results for querying down-regulated genes:
```{r, warning=FALSE, message=FALSE}
kable(head(downRegGenesResults$result[, c("term_name", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 9: Top GO-BP Results for Down-Regulated Genes")
```
<br>
Interestingly, the GO-BP results from down-regulated genes are similar to the "all genes" results with both results focusing on cellular macromolecule processes/metabolic processes.

<br>
We also can get the top Reactome results for querying down-regulated genes:
```{r, warning=FALSE, message=FALSE}
kable(head(downRegGenesResults$result[downRegGenesResults$result$source == "REAC", c("term_name", "source", "term_id", "p_value")]), type = "html", row.names = FALSE, caption = "Table 10: Top Reactome Results for Down-Regulated Genes")
```
<br>

The Reactome results from down-regulated genes are also similar to the "all genes" results with both results focusing on cell cycle and translation.

<br>

